alf/algorithms/hypernetwork_algorithm.py

```diff
@@ -0,0 +1,430 @@
+            Args for the generated parametric network
+            ====================================================================
+            input_tensor_spec (nested TensorSpec): the (nested) tensor spec of
+                the input. If nested, then ``preprocessing_combiner`` must not be
+                None.
+            conv_layer_params (tuple[tuple]): a tuple of tuples where each
+                tuple takes a format 
+                ``(filters, kernel_size, strides, padding, pooling_kernel)``,
+                where ``padding`` and ``pooling_kernel`` are optional.
+            fc_layer_params (tuple[tuple]): a tuple of tuples where each tuple
+                takes a format ``(FC layer sizes. use_bias)``, where 
+                ``use_bias`` is optional.
+            activation (nn.functional): activation used for all the layers but
+                the last layer.
+            last_layer_param (tuple): an optional tuple of the format
+                ``(size, use_bias)``, where ``use_bias`` is optional,
+                it appends an additional layer at the very end. 
+                Note that if ``last_activation`` is specified, 
+                ``last_layer_param`` has to be specified explicitly.
+            last_activation (nn.functional): activation function of the
+                additional layer specified by ``last_layer_param``. Note that if
+                ``last_layer_param`` is not None, ``last_activation`` has to be
+                specified explicitly.
+
+            Args for the generator
+            ====================================================================
+            noise_dim (int): dimension of noise
+            hidden_layers (tuple): size of hidden layers.
+            use_fc_bn (bool): whether use batnch normalization for fc layers.
+            num_particles (int): number of sampling particles
+            entropy_regularization (float): weight of entropy regularization
+
+            Args for training and testing
+            ====================================================================
+            loss_type (str): loglikelihood type for the generated functions,
+                types are [``classification``, ``regression``]
+            voting (str): types of voting results from sampled functions,
+                types are [``soft``, ``hard``]
+            par_vi (str): types of particle-based methods for variational inference,
+                types are [``svgd``, ``svgd2``, ``svgd3``, ``gfsf``]
+            optimizer (torch.optim.Optimizer): The optimizer for training.
+            logging_network (bool): whether logging the archetectures of networks.
+            logging_training (bool): whether logging loss and acc during training.
+            logging_evaluate (bool): whether logging loss and acc of evaluate.
+            config (TrainerConfig): configuration for training
+            name (str):
+        """
+        super().__init__(train_state_spec=(), name=name)
```

Comment:
A possibly better option is to use an optimizer here so that the upstream algorithm can detect other components that need to be optimized. Then pass optimizer=None to the generator constructor.

Reply:
Good point, updated.